### PR TITLE
Cleanup make_translation.py (PEP8, Python3, etc.)

### PR DESCRIPTION
**Pull request by Cabalist**
* **Please check if the PR fulfills these requirements**
- [x] The changes have been tested locally
- [x] There are no breaking changes

* **What kind of change does this PR introduce?**

In reading over the make_translations code I noticed that it had a mishmash of styles and compatibility code for Python2 but would only run under Python 3.6+.   

The following has been changed:
Remove Python2 compatibility code that would not be used
Removed dead code (`orderOutput()` and `escapeC()` were unused)
PEP8 naming and formatting
Consistent string formatting (f-strings were present so I went with those)
Add Type hinting (also present in some form prior to this PR)
Use the logging module and point it to stdout (instead of using `print()`) 
Removed assert() statements and replaced with `raise ValueError(...)` with some information on what the error is
Use list comprehension in place of `map`
Removed blanket `try:except` calls
Avoid OS specific paths in favor of Pathlib
More pythonic idioms (i.e. using `open(...)` instead of `import io; io.open(...)`

* **What is the new behavior (if this is a feature change)?**
Behaviour should be identical with the existing code.   I verified this with the following script where `make_translation_original.py` is the current HEAD file :

```python
import subprocess

all_langs = ["BG", "CS", "DA", "DE", "EN", "ES", "FI", "FR",
             "HR", "HU", "IT", "LT", "NL", "NL_BE", "NO", "PL",
             "PT", "RU", "SK", "SL", "SR_CYRL", "SR_LATN", "SV",
             "TR", "UK", "YUE_HK", "ZH_TW"]
for each in all_langs:
    subprocess.run(['python', 'make_translation.py', f'{each}', '-o', f'samples/{each}_new.c'])
    subprocess.run(['python', 'make_translation_original.py', f'{each}', '-o', f'samples/{each}_old.c'])
    with open(f'samples/{each}_old.c', 'r') as old, open(f'samples/{each}_new.c', 'r') as new:
        difference = set(old).difference(new)
        assert not difference, f"Mismatch in {each}"
```

* **Other information**:

There are a few other refactors I would like to accomplish before I am done but I wanted to get this in front of you before I made functional changes.  `get_letter_counts` is strange and does a lot of stuff internally that is just thrown away.  `write_language` and `get_font_map_and_table` can both be broken up and abstracted to be more flexible and testable in the future.   

Thank you for this project it has been incredibly helpful to me.   :)

### CODE REVIEW

**Comment by Cabalist** [comment]
I'm not sure I understand the CI failures.  It appears the Makefile can't find a target I'm not sure what would have changed to prevent these files from being created.   Any insight would be greatly appreciated. :)

**Comment by Ralim** [comment]
> I'm not sure I understand the CI failures.  It appears the Makefile can't find a target I'm not sure what would have changed to prevent these files from being created.   Any insight would be greatly appreciated. :) 

Most likely the python generator is exiting early / not creating the .cpp file correctly. Thus the source will be missing for that language and compiling will fail.

I'll try and run it locally to see what's up, but most likely is that you have changed a file name or file contents of the generated files.

1000% happy for the code changes to drop python 2 support.

**Comment by Cabalist** [comment]
@Ralim Thanks!  I thought I had tested the file output pretty thoroughly. They are byte identical with the old version. 

Perhaps the `log()` function being replaced by the stdlib `logging` may have broke something that consumed from stdout? 

I'm down for the night but if I don't see anything different I'll get the build environment setup locally. 

Thanks for the quick reply to the PR. 😃

**Comment by Cabalist** [comment]
Got it!  `fontTable.py` was renamed to match the python style (`font_table.py`)  and it got called referenced in the Makefile but not updated.

**Comment by Ralim** [comment]
Ah that makes sense.

Never liked the pep8 snake_case names 🤣

File: Translations/make_translation_test.py [hunk 1]
```diff
@@ -1,23 +1,22 @@
 #!/usr/bin/env python3
-# coding=utf-8
```
[the remark this answers is not in the document]
Comment (Cabalist) [reply Translations/make_translation_test.py:2]: That is an artifact of Python2.  It is now the default in Python3 and not recommended to include.  
> If no encoding declaration is found, the default encoding is UTF-8. 

https://docs.python.org/3/reference/lexical_analysis.html#encoding-declarations

**Comment by alvinhochun** [comment]
I don't mind some cleaning up of the script, but that means I will have to redo some of my WIP patches 😅 

While you are at it, why not also add pep8 check to the CI?

**Comment by Ralim** [comment]
> 
> 
> I don't mind some cleaning up of the script, but that means I will have to redo some of my WIP patches 😅
I've never found an excellent way to handle this sort of merge, was sort of hoping we could rebase your patches on top or are we going to have conflict hell?
> 
> While you are at it, why not also add pep8 check to the CI?
I would love to see pep8+black added. I'm happy to add these after these changes though; I do have them in other repo's so should be easy to port across.

**Comment by alvinhochun** [comment]
> 
> 
> > I don't mind some cleaning up of the script, but that means I will have to redo some of my WIP patches 😅
> 
> I've never found an excellent way to handle this sort of merge, was sort of hoping we could rebase your patches on top or are we going to have conflict hell?

No biggie, I can just rewrite them. I am not even sure if I will make PRs out of them.

**Review by Ralim — APPROVED** [review]

**Comment by Ralim** [comment]
@Cabalist let me know if you want to add pep8 and black to the ci steps or if there is more you would like to do :)

**Comment by alvinhochun** [comment]
Another thing is that I'm not sure the benefits of transforming some of the `assert`s to explicitly raising `ValueError`s, because most of the asserts (with some exceptions) are just sanity checks. I don't really mind either way, but don't expect me to intentionally avoid using `assert`s in future PRs 😅

**Comment by Cabalist** [comment]
@Ralim Go ahead.   I fully support `black` cause it does a great job of ending any style arguments.  :) 

@alvinhochun 
**_\<begin style argument>_** 
`assert`s in production code are generally frowned on.   Rule of thumb is that they should only be used in test cases.   A more targeted `Exception` has a bunch of benefits including communicating to the user what's happened and allowing programs that use the code to catch and handle specific cases instead of just taking the whole program down.   `assert`s in non-test code will often raising linting errors with `flake8` and other linters by default for these reasons.

[This stackoverflow answer](https://stackoverflow.com/a/945135/1058858) is pretty succinct on the issue:
>Asserts should be used to test conditions that should never happen. The purpose is to crash early in the case of a corrupt program state.
>
>Exceptions should be used for errors that can conceivably happen, and you should almost always create your own Exception classes.
>
>For example, if you're writing a function to read from a configuration file into a dict, improper formatting in the file should raise a ConfigurationSyntaxError, while you can assert that you're not about to return None.
>
>In your example, if x is a value set via a user interface or from an external source, an exception is best.
>
>If x is only set by your own code in the same program, go with an assertion.

 **_\<end style argument>_**

**Comment by Ralim** [comment]
My personal preference is to assert and hard fail catestrophically in this case since this python code is never run in _production_ but is used as a bootstrap before the compile stage. Thus we want any issues at all in the python code to explode and grind the CI to a halt.

That said, given the number of things in the PEP world I'm not a fan of, I've sort of given up on that argument 🙃

**Comment by alvinhochun** [comment]
@Cabalist 

> Asserts should be used to test conditions that should never happen. The purpose is to crash early in the case of a corrupt program state.

This is exactly what I used those asserts for. Not to mention that the script is for internal use and the "user" is whoever wrote and maintain the scripts, and I feel that the exception/error message doesn't actually say anything more than what the code itself reads, so they just feel redundant, but maybe that's just me.

**Pull request merged**